charmdir.go

```diff
@@ -454,19 +453,84 @@ func (m NopLogger) Infof(message string, args ...interface{}) {
 
 }
 
+type vcsCMD struct {
+	vcsType       string
+	args          []string
+	usesTypeCheck func(charmPath string) bool
+}
+
+func (v *vcsCMD) commonErrHandler(err error, charmPath string) error {
+	return errors.Errorf("%q version string generation failed : "+
+		"%v\nThis means that the charm version won't show in juju status. Charm path %q", v.vcsType, err, charmPath)
+}
+
+// The first check checks for the easy case of the current charmdir has a git folder.
+// There can be cases when the charmdir actually uses git and is just a subdir, thus the below check
+func usesGit(charmPath string) bool {
+	if _, err := os.Stat(filepath.Join(charmPath, ".git")); err == nil {
+		return true
+	}
+	args := []string{"rev-parse", "--is-inside-work-tree"}
+	execCmd := exec.Command("git", args...)
+	execCmd.Dir = charmPath
+	if out, err := execCmd.Output(); err == nil {
+		logger.Errorf("%q", out)
+		return true
+	}
+	return false
+}
+
+func usesBzr(charmPath string) bool {
+	if _, err := os.Stat(filepath.Join(charmPath, ".bzr")); err == nil {
+		return true
+	}
+	return false
+}
+
+func usesHg(charmPath string) bool {
+	if _, err := os.Stat(filepath.Join(charmPath, ".hg")); err == nil {
+		return true
+	}
+	return false
+}
+
 // MaybeGenerateVersionString generates charm version string.
+// We want to know whether parent folders use one of these vcs, that's why we try to execute each one of them
 // The second return value is the detected vcs type.
 func (dir *CharmDir) MaybeGenerateVersionString(logger Logger) (string, string, error) {
-	var cmdArgs []string
+	vcsStrategies := make(map[string]vcsCMD)
+
 	versionFileVersionType := "versionFile"
-	// Verify that it is revision control directory.
-	if _, err := os.Stat(filepath.Join(dir.Path, ".hg")); err == nil {
-		cmdArgs = []string{"hg", "id", "-n"}
-	} else if _, err = os.Stat(filepath.Join(dir.Path, ".bzr")); err == nil {
-		cmdArgs = []string{"bzr", "version-info"}
-	} else if _, err = os.Stat(filepath.Join(dir.Path, ".git")); err == nil {
-		cmdArgs = []string{"git", "describe", "--dirty", "--always"}
-	} else if file, err := os.Open(dir.join("version")); err == nil {
+	mercurialStrategy := vcsCMD{"hg", []string{"id", "-n"}, usesHg}
+	bazaarStrategy := vcsCMD{"bzr", []string{"version-info"}, usesBzr}
+	gitStrategy := vcsCMD{"git", []string{"describe", "--dirty", "--always"}, usesGit}
+
+	vcsStrategies["hg"] = mercurialStrategy
+	vcsStrategies["git"] = gitStrategy
+	vcsStrategies["bzr"] = bazaarStrategy
+
+	// Nowadays most vcs used are git, we want to make sure that git is the first one we test
+	vcsOrder := [...]string{"git", "hg", "bzr"}
+
+	for _, vcsType := range vcsOrder {
+		vcsCmd := vcsStrategies[vcsType]
+		if !vcsCmd.usesTypeCheck(dir.Path) {
```

Comment:
Can we add a comment that if all vcs strategies fail, then it will fall back to the version check below the for loop.